pytorch_lightning/strategies/collaborative.py

```diff
@@ -0,0 +1,529 @@
+                "requires a `scheduler_fn` to be passed to the strategy if a scheduler is being used "
+                "(this is because the optimizer is re-created within Hivemind)."
+            )
+
+        scheduler = self._scheduler_fn if self._require_scheduler_fn else None
+        params = optimizer.param_groups if self._require_scheduler_fn else None
+        optimizer = type(optimizer) if self._require_scheduler_fn else optimizer
+        opt = hivemind.Optimizer(
+            dht=self.dht,
+            run_id=self._run_id,
+            params=params,
+            optimizer=optimizer,
+            scheduler=scheduler,
+            target_batch_size=self._target_batch_size,
+            batch_size_per_step=self._batch_size,
+            **self._optimizer_kwargs,
+        )
+
+        if not self._scheduler_fn:
+            self._wrap_schedulers(opt)
+        opt.load_state_from_peers()
+        self.optimizers = [opt]
+        self._opt = opt
+
+        if self._reuse_grad_buffers:
+            assert self.lightning_module is not None
+            self._optimizer_zero_grad_original = self.lightning_module.optimizer_zero_grad
+            self._disable_zero_grad()
+
+    def _disable_zero_grad(self) -> None:
+        lightning_module = self.lightning_module
+        if is_overridden("optimizer_zero_grad", lightning_module):
+            assert lightning_module is not None  # `is_overridden` returns False otherwise
+            rank_zero_warn(
+                "You have overridden `optimizer_zero_grad` which will be disabled."
+                " When `CollaborativeStrategy(reuse_grad_buffers=True)`, the optimizer cannot call zero grad,"
+                " as this would delete the gradients before they are averaged."
+            )
+        assert lightning_module is not None
+        lightning_module.optimizer_zero_grad = None  # type: ignore[assignment]
+
+    def _wrap_schedulers(self, opt: "hivemind.Optimizer") -> None:
+        # wrap schedulers so that they only update when the hivemind optimizer updates
+        for scheduler_config in self.lr_scheduler_configs:
+            scheduler = scheduler_config.scheduler
+            if isinstance(scheduler, ReduceLROnPlateau):
+                raise ValueError(
+                    f"The `ReduceLROnPlateau` scheduler is not currently supported with `{self.__class__.__name__}`."
+                )
+            scheduler_config.scheduler = HiveMindScheduler(
+                optimizer=opt,
+                scheduler=scheduler,
+            )
+
+    def on_train_batch_start(self, batch: Any, batch_idx: int, dataloader_idx: int = 0) -> None:
+        if not self._hivemind_initialized:
+            self._hivemind_initialized = True
+            # todo (sean): we could technically support a dynamic batch size by inferring each step
+            # and passing it to the ``hivemind.Optimizer``.
+            if self._batch_size is None:
```

Comment:
what if batch_size differs for validation/test.. will it consider it the same for all?

Reply:
This only affects training, no sync during test/validation